qiskit_ibm_runtime/utils/utils.py

```diff
@@ -116,6 +120,113 @@ def is_isa_circuit(circuit: QuantumCircuit, target: Target) -> str:
     return _is_isa_circuit_helper(circuit, target, qubit_map)
 
 
+def _is_rzz_pub_helper(circuit: QuantumCircuit) -> Union[str, Set[Parameter]]:
+    """
+    For rzz gates:
+    - Verify that numeric angles are in the range [0, pi/2]
+    - Collect parameterized angles
+
+    Returns one of the following:
+    - A string, containing an error message, if a numeric angle is outside of the range [0, pi/2]
+    - A list of names of all the parameters that participate in an rzz gate
+
+    Note: we check for parametrized rzz gates inside control flow operation, although fractional
+    gates are actually impossible in combination with dynamic circuits. This is in order to remain
+    correct if this restriction is removed at some point.
+    """
+    angle_params = set()
+
+    for instruction in circuit.data:
+        operation = instruction.operation
+
+        # rzz gate is calibrated only for the range [0, pi/2].
+        # We allow an angle value of a bit more than pi/2, to compensate floating point rounding
+        # errors (beyond pi/2 does not trigger an error down the stack, only may become less
+        # accurate).
+        if operation.name == "rzz":
+            angle = instruction.operation.params[0]
+            if isinstance(angle, Parameter):
+                angle_params.add(angle.name)
+            elif not isinstance(angle, ParameterExpression) and (
+                angle < 0.0 or angle > 1.001 * np.pi / 2
```

Comment:
There is an asymmetry here in that rounding errors are compensated on the upper end, but not the lower end. for example, `-1e-12` is invalid whereas `np.pi/2 + 1e-12` is valid.

I'm also  a bit surprised that the level is at 1e-3, very far away from machine epsilon. For example, is there a problem moving this to 1e-10 which is still 100x machine epsilon?

```suggestion
                not -EPSILON < angle < np.pi / 2 + EPSILON
```

Reply:
This mirrors line 82 above. I think I contributed to the original choice of 1.001. The hardware is optimized for pi/2 because the main focus is the cz gate. It doesn't blow up if you go just past pi/2 but you don't want to go much past it (not calibrated and will eventually hit some hardware/physical limit). So it is not critical that the limit be tight on pi/2 but whatever is likely not to run into rounding issues should be okay.

Reply:
Then, what's the decision, for both ends?

Reply:
Once there is a decision, I'll of course fix (if needed) not only here, but also in several other places where this check takes place (I've spotted that this check appears in three different places, maybe I'll write a tiny function for it).

Reply:
I am okay with 1e-10. I think we only document pi/2 as the limit and here we are just trying to avoid being too strict where in practice the user won't notice the difference.

There was also the point about the asymmetry with negative angles. I am not sure what to do about that right now. The backend will error right now if it gets a negative number. We could change the backend to round small negative numbers to 0 if we wanted symmetry of behavior for the bounds.

Reply:
Based on this conversation, I'm keeping the current behavior for the negative angles. Let me know if it's a problem.

Reply:
All conversations must be resolved before merging. Can we mark this as resolved @ihincks? I think the remaining options would be to round up small negative angles/parameters to 0 (a change of scope since the validation code only checks without modifying right now) or to change the backend code to accept small negative angles (out of scope here for now).